tidy3d/plugins/design/design.py

```diff
@@ -102,22 +128,190 @@ def run(self, function: Callable, **kwargs) -> Result:
+        """Get function that tries to use batch processing on a set of arguments."""
+
+        class Pre_Post_Handler:
+            def __init__(self):
+                self.sim_counter = 0
+                self.sim_names = []
+
+            def _fn_combined(self, args_list):
+                """Compute fn_pre and fn_post functions and capture other outputs."""
+                sim_dict = {str(idx): fn_pre(**arg_list) for idx, arg_list in enumerate(args_list)}
+                data, sim_names = fn_mid(sim_dict, self.sim_counter)
+                self.sim_names.extend(sim_names)
+                self.sim_counter += len(sim_names)
+                post_out = [fn_post(val[1]) for val in data.items()]
+                return post_out
+
+            def evaluate(self, args_list: list[tuple[Any, ...]]) -> list[Any]:
+                """Put together into one pipeline."""
+                out = self._fn_combined(args_list)
+                return out
+
+        handler = Pre_Post_Handler()
+
+        return handler
+
+    def _fn_mid(
+        self, pre_out: dict[int, Any], sim_counter: int
+    ) -> Union[dict[int, Any], BatchData]:
+        """A function of the output of ``fn_pre`` that gives the input to ``fn_post``."""
+
+        # NOTE: Need to check that pre_out has string keys!
+
+        # Keep copy of original to use if no tidy3d simulation required
+        original_pre_out = pre_out.copy()
+
+        # Convert list input to dict of dict before passing through checks
+        was_list = False
+        if all(isinstance(sim, list) for sim in pre_out.values()):
+            pre_out = {
+                str(idx): {str(sub_idx): sub_list for sub_idx, sub_list in enumerate(sim_list)}
+                for idx, sim_list in enumerate(pre_out.values())
+            }
+            was_list = True
+
+        def _find_and_map(
+            search_dict: dict,
+            search_type: Any,
+            output_dict: dict,
+            naming_dict: dict,
+            previous_key: str = "",
+        ):
+            """Recursively search for search_type objects within a dictionary."""
+            current_key = previous_key
+            for key, value in search_dict.items():
+                if len(previous_key) == 0:
+                    latest_key = str(key)
+                else:
+                    latest_key = f"{current_key}_{key}"
+
+                if isinstance(value, dict):
+                    _find_and_map(value, search_type, output_dict, naming_dict, latest_key)
+                elif isinstance(value, search_type):
+                    output_dict[latest_key] = value
+                    naming_dict[latest_key] = key
+
+        simulations = {}
+        batches = {}
+        naming_keys = {}
+
+        _find_and_map(pre_out, Simulation, simulations, naming_keys)
+        _find_and_map(pre_out, Batch, batches, naming_keys)
+
+        # Exit fn_mid here if td computation is required
+        if len(simulations) == 0 and len(batches) == 0:
```

Comment:
This is here not so much because the next section can't deal with empty batches or simulations, but because the return statement needs to be different. Having this saves converting back to the original format of pre_out and handling no named_sims data.